<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Results on mixFine runs on GTEx V8 (subsampling analysis)</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="site_libs/anchor-sections-1.0/anchor-sections.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">mixqtl-pipeline</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/liangyy/mixqtl-pipeline">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Results on mixFine runs on GTEx V8 (subsampling analysis)</h1>

</div>


<pre class="r"><code>rm(list = ls())
library(ggplot2)
theme_set(theme_bw(base_size = 12))
library(dplyr)
library(reshape2)
library(stringr)
library(data.table)
options(datatable.fread.datatable = FALSE)
options(stringsAsFactors = FALSE)
source(&#39;../code/rlib_analysis.R&#39;)
datadir = &#39;/Users/yanyul/Desktop/mixqtl-pipeline-results/&#39;  # gtex_v8-results/
cbPalette &lt;- c(&#39;ascQTL&#39; = &quot;#999999&quot;, &#39;mixFine&#39; = &quot;#E69F00&quot;, &#39;trcQTL&#39; = &quot;#56B4E9&quot;, &#39;vanilla&#39; = &quot;#009E73&quot;, &#39;standard approach&#39; = &quot;#009E73&quot;, &quot;#F0E442&quot;, &quot;#0072B2&quot;, &quot;#D55E00&quot;, &quot;#CC79A7&quot;)
source(&#39;https://gist.githubusercontent.com/liangyy/43912b3ecab5d10c89f9d4b2669871c9/raw/8151c6fe70e3d4ee43d9ce340ecc0eb65172e616/my_ggplot_theme.R&#39;)
# th0 = th
# th$panel.border = element_rect(colour = th$axis.line$colour)

parse_name = function(ee) {
  ee = basename(ee)
  b = as.numeric(str_remove(str_extract(ee, &#39;batch([0-9]+)&#39;), &#39;batch&#39;))
  f = as.numeric(str_remove(str_extract(ee, &#39;fraction([0-9\\.]+)&#39;), &#39;fraction&#39;))
  cc = str_remove(str_extract(ee, &#39;result_([a-z]+)&#39;), &#39;result_&#39;)
  mm = str_remove(str_extract(ee, paste0(&#39;result_&#39;, cc, &#39;-[a-z]+&#39;)), paste0(&#39;result_&#39;, cc, &#39;-&#39;))
  return(list(batch = b, fraction = f, type = cc, method = mm))
}

expression_level = read.table(&#39;~/Desktop/mixqtl-pipeline-results/misc_data/gtex-v8-genes-passed-qc-with-median-trc.txt&#39;, header = T, stringsAsFactors = F)</code></pre>
<div id="load-result" class="section level1">
<h1><span class="header-section-number">1</span> Load result</h1>
<pre class="r"><code>method_df = data.frame(method = c(&#39;mixfine&#39;, &#39;nefine&#39;), tag = c(&#39;mixFine&#39;, &#39;standard approach&#39;))
fractions = 3 : 9 / 10
batch = 1 : 10

# load results from full data
df_full = list()
df_full_g = list()

for(m in 1 : nrow(method_df)) {
  method = method_df$method[m]
  tag = method_df$tag[m]
  d = readRDS(paste0(datadir, &#39;mixfine/Whole_Blood/pack/GTExV8_WholeBlood__result_pip-&#39;, method, &#39;.rds&#39;))
  df_full[[length(df_full) + 1]] = d$pip %&gt;% mutate(method = method, batch = NA, fraction = 1)
  df_full_g[[length(df_full_g) + 1]] = data.frame(gene = d$gene, method = method, batch = NA, fraction = 1)
}
df_full = do.call(rbind, df_full)
df_full_g = do.call(rbind, df_full_g)

# load results from full data
list_pip = list()
list_g = list()
for(m in 1 : nrow(method_df)) {
  method = method_df$method[m]
  tag = method_df$tag[m]
  for(frac in fractions) {
    for(bat in batch) {
      e = Sys.glob(paste0(datadir, &#39;mixfine-subsample/gene_list_whole_blood_for_subsampling_with_qc0*/Whole_Blood/pack/mypack-08-10-20__indiv_subset-indiv_subset__fraction&#39;, frac, &#39;__batch&#39;, bat, &#39;-result_pip-&#39;, method, &#39;.rds&#39;))
      for(ee in e) {
        d = readRDS(ee)
      list_pip[[length(list_pip) + 1]] = d$pip %&gt;% mutate(method = method, batch = bat, fraction = frac)
      list_g[[length(list_g) + 1]] = data.frame(gene = d$gene, method = method, batch = bat, fraction = frac)
      }
    }
  }
}
df_subsample = do.call(rbind, list_pip)
df_subsample_g = do.call(rbind, list_g)

# 
df_subsample_g_complete = df_subsample_g %&gt;% group_by(method, gene) %&gt;% summarize(n_success = n()) %&gt;% ungroup() %&gt;% filter(n_success == 10 * 7) %&gt;% group_by(gene) %&gt;% summarize(n_method = n()) %&gt;% ungroup() %&gt;% filter(n_method == 2)</code></pre>
<pre><code>## `summarise()` regrouping output by &#39;method&#39; (override with `.groups` argument)</code></pre>
<pre><code>## `summarise()` ungrouping output (override with `.groups` argument)</code></pre>
<pre class="r"><code>df_full_g_complete = df_full_g %&gt;% group_by(gene) %&gt;% summarize(n_success = n()) %&gt;% ungroup() %&gt;% filter(n_success == 2)</code></pre>
<pre><code>## `summarise()` ungrouping output (override with `.groups` argument)</code></pre>
<pre class="r"><code>gene_complete_all = df_subsample_g_complete %&gt;% filter(gene %in% df_full_g_complete$gene)

# extract the genes with complete results (no failed runs)
df_subsample_complete = df_subsample %&gt;% filter(gene %in% gene_complete_all$gene) 
df_full_complete = df_full %&gt;% filter(gene %in% gene_complete_all$gene)
df_both_complete = rbind(df_full_complete, df_subsample_complete)

# # filter out low asc genes
# df_subsample_complete = df_subsample_complete %&gt;% filter(gene %in% expression_level$gene[expression_level$n_good_asc &gt; 30 / 3 * 10]) 
# df_full_complete = df_full_complete %&gt;% filter(gene %in% expression_level$gene[expression_level$n_good_asc &gt; 30 / 3 * 10]) 
# df_both_complete = df_both_complete %&gt;% filter(gene %in% expression_level$gene[expression_level$n_good_asc &gt; 30 / 3 * 10]) </code></pre>
</div>
<div id="plot" class="section level1">
<h1><span class="header-section-number">2</span> Plot</h1>
<div id="number-of-95-credible-set-identified-in-total" class="section level2">
<h2><span class="header-section-number">2.1</span> Number of 95% credible set identified in total</h2>
<pre class="r"><code>ncs_total = df_subsample_complete %&gt;% group_by(gene, method, fraction) %&gt;% summarize(ncs = length(unique(cs))) %&gt;% ungroup()</code></pre>
<pre><code>## `summarise()` regrouping output by &#39;gene&#39;, &#39;method&#39; (override with `.groups` argument)</code></pre>
<pre class="r"><code>g = ncs_total %&gt;% dcast(fraction + gene ~ method, value.var = &#39;ncs&#39;, fill = 0)</code></pre>
<pre><code>## Warning in dcast(., fraction + gene ~ method, value.var = &quot;ncs&quot;, fill = 0): The dcast generic in data.table has been passed a
## tbl_df and will attempt to redirect to the reshape2::dcast; please note that reshape2 is deprecated, and this redirection is now
## deprecated as well. Please do this redirection yourself like reshape2::dcast(.). In the next version, this warning will become an
## error.</code></pre>
<pre class="r"><code>g %&gt;% ggplot() + geom_point(aes(x = nefine, y = mixfine)) + facet_wrap(~fraction)</code></pre>
<p><img src="gtex_v8_finemap_subsample_files/figure-html/unnamed-chunk-25-1.png" width="672" /></p>
</div>
<div id="concensus-snps" class="section level2">
<h2><span class="header-section-number">2.2</span> Concensus SNPs</h2>
<pre class="r"><code># extract variant-gene pairs with PIP &gt; 0.5 from full set in both mixfine and nefine
# plot the fraction of them identified in sub sampled sets
consensus_eqtl = df_full_complete %&gt;% filter(variable_prob &gt; 0.5) %&gt;% group_by(variant_id, gene) %&gt;% summarize(n_method = n()) %&gt;% ungroup() %&gt;% filter(n_method == 2)</code></pre>
<pre><code>## `summarise()` regrouping output by &#39;variant_id&#39; (override with `.groups` argument)</code></pre>
<pre class="r"><code>df_subsample_with_consensus = df_subsample_complete %&gt;% filter(variable_prob &gt; 0) %&gt;% mutate(is_consensus = paste(variant_id, gene) %in% paste(consensus_eqtl$variant_id, consensus_eqtl$gene))
df_subsample_with_consensus_summary = df_subsample_with_consensus %&gt;% group_by(method, fraction, batch) %&gt;% summarize(fraction_of_detection = sum(is_consensus) / nrow(consensus_eqtl))</code></pre>
<pre><code>## `summarise()` regrouping output by &#39;method&#39;, &#39;fraction&#39; (override with `.groups` argument)</code></pre>
<pre class="r"><code>df_subsample_with_consensus_summary_flat = df_subsample_with_consensus_summary %&gt;% dcast(fraction + batch ~ method, value.var = &#39;fraction_of_detection&#39;)</code></pre>
<pre><code>## Warning in dcast(., fraction + batch ~ method, value.var = &quot;fraction_of_detection&quot;): The dcast generic in data.table has been
## passed a grouped_df and will attempt to redirect to the reshape2::dcast; please note that reshape2 is deprecated, and this
## redirection is now deprecated as well. Please do this redirection yourself like reshape2::dcast(.). In the next version, this
## warning will become an error.</code></pre>
<pre class="r"><code>df_subsample_with_consensus_summary_flat %&gt;% ggplot() + geom_point(aes(x = mixfine, y = nefine)) + geom_abline(slope = 1, intercept = 0)</code></pre>
<p><img src="gtex_v8_finemap_subsample_files/figure-html/unnamed-chunk-26-1.png" width="672" /></p>
<pre class="r"><code>df_subsample_with_consensus_summary %&gt;% ggplot() + geom_boxplot(aes(x = as.factor(fraction), y = fraction_of_detection, color = method)) + scale_color_manual(values = cbPalette)</code></pre>
<p><img src="gtex_v8_finemap_subsample_files/figure-html/unnamed-chunk-26-2.png" width="672" /></p>
<pre class="r"><code># for saving a nice plot
temp = df_subsample_with_consensus_summary
temp$method[temp$method == &#39;mixfine&#39;] = &#39;mixFine&#39;
temp$method[temp$method == &#39;nefine&#39;] = &#39;standard approach&#39;
p = temp %&gt;% ggplot() + geom_violin(aes(x = as.factor(fraction), y = fraction_of_detection, fill = method), alpha = .5, position = position_dodge(1)) + 
  # geom_boxplot(aes(x = as.factor(fraction), y = fraction_of_detection, fill = method), alpha = 1, width = .2, position = position_dodge(1)) + 
  scale_fill_manual(values = cbPalette) + xlab(&#39;Subsampling fraction&#39;) + ylab(&#39;Power (detection rate)&#39;) + theme(legend.position = c(0.8, 0.2), legend.title = element_blank()) + geom_jitter(aes(x = as.factor(fraction), y = fraction_of_detection, color = method), position = position_dodge(1), size = 0.7, alpha = 1) # + ggtitle(&#39;Consensus variants \n (PIP &gt; 0.5 in both) \n being detected&#39;)
p = p + th + theme(aspect.ratio = 1)
p</code></pre>
<p><img src="gtex_v8_finemap_subsample_files/figure-html/unnamed-chunk-26-3.png" width="672" /></p>
<pre class="r"><code>ggsave(&#39;../output/gtex-v8-pipeline-eqtl-finemapping-concensus-snp-with-gene-qc.png&#39;, p, height = 4, width = 4)
ggsave(&#39;../output/gtex-v8-pipeline-eqtl-finemapping-concensus-snp-with-gene-qc.pdf&#39;, p, height = 4, width = 4)</code></pre>
</div>
<div id="concensus-css" class="section level2">
<h2><span class="header-section-number">2.3</span> Concensus CSs</h2>
<pre class="r"><code># get variants in Consensus CS&#39;s (sharing the same variant)
sub = df_full %&gt;% filter(cs &gt; 0) %&gt;% group_by(gene, variant_id) %&gt;% summarize(nmethod = length(unique(method))) %&gt;% ungroup() %&gt;% filter(nmethod &gt; 1)</code></pre>
<pre><code>## `summarise()` regrouping output by &#39;gene&#39; (override with `.groups` argument)</code></pre>
<pre class="r"><code># get Consensus top variant of Consensus CS
sub2 = df_full %&gt;% filter(paste(gene, variant_id) %in% paste(sub$gene, sub$variant_id))
sub2 = sub2 %&gt;% group_by(gene, method, cs) %&gt;% summarize(top_var = variant_id[which.max(variable_prob)])</code></pre>
<pre><code>## `summarise()` regrouping output by &#39;gene&#39;, &#39;method&#39; (override with `.groups` argument)</code></pre>
<pre class="r"><code>sub2 = sub2 %&gt;% group_by(gene, top_var) %&gt;% summarize(n = n()) %&gt;% ungroup() %&gt;% filter(n == 2)</code></pre>
<pre><code>## `summarise()` regrouping output by &#39;gene&#39; (override with `.groups` argument)</code></pre>
<pre class="r"><code># get cs size in subsample
sub_cs_size_sub = df_subsample %&gt;% group_by(gene, cs, method, fraction, batch) %&gt;% summarize(cs_size = n())</code></pre>
<pre><code>## `summarise()` regrouping output by &#39;gene&#39;, &#39;cs&#39;, &#39;method&#39;, &#39;fraction&#39; (override with `.groups` argument)</code></pre>
<pre class="r"><code># get cs with top variant in subsample
sub_cs = df_subsample %&gt;% filter(paste(gene, variant_id) %in% paste(sub2$gene, sub2$top_var))
# annotate extracted cs with cs size
sub_cs = left_join(sub_cs, sub_cs_size_sub, by = c(&#39;gene&#39;, &#39;cs&#39;, &#39;batch&#39;, &#39;fraction&#39;, &#39;method&#39;))
# flat to compare mixfine vs nefine
sub_flat = sub_cs %&gt;% dcast(gene + variant_id + batch + fraction ~ method, value.var = &#39;cs_size&#39;)</code></pre>
<pre><code>## Warning in dcast(., gene + variant_id + batch + fraction ~ method, value.var = &quot;cs_size&quot;): The dcast generic in data.table has
## been passed a data.frame and will attempt to redirect to the reshape2::dcast; please note that reshape2 is deprecated, and this
## redirection is now deprecated as well. Please do this redirection yourself like reshape2::dcast(.). In the next version, this
## warning will become an error.</code></pre>
<pre class="r"><code># plot: for cs&#39;s detected in both mixfine and nefine
th2 = th
th2$panel.border = element_rect(colour = th2$axis.line$colour)
p = sub_flat %&gt;% filter(!is.na(mixfine), !is.na(nefine)) %&gt;% filter(fraction != 0.3) %&gt;% ggplot(aes(y = mixfine, x = nefine)) + geom_abline(slope = 1, intercept = 0, color = &#39;lightgray&#39;) + geom_point(alpha = .5) + facet_wrap(~fraction, ncol = 2) + ylab(&#39;CS size in mixFine&#39;) + xlab(&#39;CS size in standard approach&#39;) + coord_equal(xlim = c(0, 300), y = c(0, 300)) # + ggtitle(&#39;Size of Consensus CS \n being detected in both&#39;)
p = p + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5)) # + coord_cartesian(xlim = c(0, 300), y = c(0, 300))
p = p + th2
p</code></pre>
<p><img src="gtex_v8_finemap_subsample_files/figure-html/unnamed-chunk-27-1.png" width="672" /></p>
<pre class="r"><code>ggsave(&#39;../output/gtex-v8-pipeline-eqtl-finemapping-cs-size-gene-qc.png&#39;, p, height = 5, width = 4)

# plot: counting how many cs&#39;s are detected
sub_sum = sub_cs %&gt;% group_by(method, fraction, batch) %&gt;% summarize(ntimes_detected = n()) </code></pre>
<pre><code>## `summarise()` regrouping output by &#39;method&#39;, &#39;fraction&#39; (override with `.groups` argument)</code></pre>
<pre class="r"><code>sub_sum_flat = sub_sum %&gt;% dcast(fraction + batch ~ method, value.var = &#39;ntimes_detected&#39;)</code></pre>
<pre><code>## Warning in dcast(., fraction + batch ~ method, value.var = &quot;ntimes_detected&quot;): The dcast generic in data.table has been passed a
## grouped_df and will attempt to redirect to the reshape2::dcast; please note that reshape2 is deprecated, and this redirection is
## now deprecated as well. Please do this redirection yourself like reshape2::dcast(.). In the next version, this warning will become
## an error.</code></pre>
<pre class="r"><code>p = sub_sum_flat %&gt;% ggplot() + geom_abline(slope = 1, intercept = 0, color = &#39;lightgray&#39;) + geom_point(aes(y = mixfine / nrow(sub2), x = nefine / nrow(sub2))) + ylab(&#39;Detection rate in mixFine&#39;) + xlab(&#39;Detection rate in standard approach&#39;) + coord_equal() # + ggtitle(&#39;Consensus CS \n (CS with same top variant) \n being detected&#39;)
p = p + th
p</code></pre>
<p><img src="gtex_v8_finemap_subsample_files/figure-html/unnamed-chunk-27-2.png" width="672" /></p>
<pre class="r"><code>ggsave(&#39;../output/gtex-v8-pipeline-eqtl-finemapping-concensus-cs-gene-qc.png&#39;, p, height = 4, width = 4)

sub_flat %&gt;% filter(!is.na(mixfine), !is.na(nefine)) %&gt;% summarize(mean_mixfine = mean(mixfine, na.rm = T), mean_nefine = mean(nefine, na.rm = T))</code></pre>
<pre><code>##   mean_mixfine mean_nefine
## 1     9.491445    14.56412</code></pre>
</div>
</div>
<div id="stratifying-by-gene-expression-level" class="section level1">
<h1><span class="header-section-number">3</span> Stratifying by gene expression level</h1>
<pre class="r"><code>cutoffs = quantile(expression_level$median_trc, probs = c(1/3, 2/3))
# df_subsample_with_consensus$median_trc = expression_level$median_trc[match(df_subsample_with_consensus$gene, expression_level$gene)]
df_subsample_with_consensus = df_subsample_complete %&gt;% mutate(is_consensus = paste(variant_id, gene) %in% paste(consensus_eqtl$variant_id, consensus_eqtl$gene))
df_subsample_with_consensus$expression_level = &#39;high&#39;
df_subsample_with_consensus$expression_level[df_subsample_with_consensus$gene %in% expression_level$gene[expression_level$median_trc &lt; cutoffs[1]]] = &#39;low&#39;
df_subsample_with_consensus$expression_level[df_subsample_with_consensus$gene %in% expression_level$gene[expression_level$median_trc &lt; cutoffs[2] &amp; expression_level$median_trc &gt;= cutoffs[1]]] = &#39;middle&#39;


# consensus_eqtl$median_trc = expression_level$median_trc[match(consensus_eqtl$gene, expression_level$gene)]
consensus_eqtl$expression_level = &#39;high&#39;
consensus_eqtl$expression_level[consensus_eqtl$gene %in% expression_level$gene[expression_level$median_trc &lt; cutoffs[1]]] = &#39;low&#39;
consensus_eqtl$expression_level[consensus_eqtl$gene %in% expression_level$gene[expression_level$median_trc &lt; cutoffs[2] &amp; expression_level$median_trc &gt;= cutoffs[1]]] = &#39;middle&#39;
n_consensus_eqtl = consensus_eqtl %&gt;% group_by(expression_level) %&gt;% summarize(n = n()) %&gt;% ungroup()</code></pre>
<pre><code>## `summarise()` ungrouping output (override with `.groups` argument)</code></pre>
<pre class="r"><code>df_subsample_with_consensus_summary = df_subsample_with_consensus %&gt;% group_by(method, fraction, batch, expression_level) %&gt;% summarize(num_of_detection = sum(is_consensus))</code></pre>
<pre><code>## `summarise()` regrouping output by &#39;method&#39;, &#39;fraction&#39;, &#39;batch&#39; (override with `.groups` argument)</code></pre>
<pre class="r"><code>df_subsample_with_consensus_summary = inner_join(df_subsample_with_consensus_summary, n_consensus_eqtl, by = &#39;expression_level&#39;)
df_subsample_with_consensus_summary = df_subsample_with_consensus_summary %&gt;% mutate(fraction_of_detection = num_of_detection / n)
df_subsample_with_consensus_summary$expression_level = factor(df_subsample_with_consensus_summary$expression_level, levels = c(&#39;high&#39;, &#39;middle&#39;, &#39;low&#39;))

df_subsample_with_consensus_summary_flat = df_subsample_with_consensus_summary %&gt;% dcast(fraction + batch + expression_level ~ method, value.var = &#39;fraction_of_detection&#39;)</code></pre>
<pre><code>## Warning in dcast(., fraction + batch + expression_level ~ method, value.var = &quot;fraction_of_detection&quot;): The dcast generic
## in data.table has been passed a grouped_df and will attempt to redirect to the reshape2::dcast; please note that reshape2 is
## deprecated, and this redirection is now deprecated as well. Please do this redirection yourself like reshape2::dcast(.). In the
## next version, this warning will become an error.</code></pre>
<pre class="r"><code>df_subsample_with_consensus_summary_flat %&gt;% ggplot() + geom_point(aes(x = mixfine, y = nefine)) + geom_abline(slope = 1, intercept = 0) + facet_wrap(~expression_level)</code></pre>
<p><img src="gtex_v8_finemap_subsample_files/figure-html/unnamed-chunk-28-1.png" width="672" /></p>
<pre class="r"><code>df_subsample_with_consensus_summary %&gt;% ggplot() + geom_boxplot(aes(x = as.factor(fraction), y = fraction_of_detection, color = method)) + facet_wrap(~expression_level) # + scale_color_manual(values = cbPalette)</code></pre>
<p><img src="gtex_v8_finemap_subsample_files/figure-html/unnamed-chunk-28-2.png" width="672" /></p>
<pre class="r"><code># for saving a nice plot
temp = df_subsample_with_consensus_summary
temp$method[temp$method == &#39;mixfine&#39;] = &#39;mixFine&#39;
temp$method[temp$method == &#39;nefine&#39;] = &#39;standard approach&#39;
p = temp %&gt;% ggplot() + geom_violin(aes(x = as.factor(fraction), y = fraction_of_detection, fill = method), alpha = .5, position = position_dodge(1)) + 
  # geom_boxplot(aes(x = as.factor(fraction), y = fraction_of_detection, fill = method), alpha = 1, width = .2, position = position_dodge(1)) + 
  scale_fill_manual(values = cbPalette) + xlab(&#39;Subsampling fraction&#39;) + ylab(&#39;Power (detection rate)&#39;) + theme(legend.position = &#39;bottom&#39;, legend.title = element_blank()) + facet_wrap(~expression_level) + geom_point(aes(x = as.factor(fraction), y = fraction_of_detection, color = method), position = position_dodge(1), size = 0.7, alpha = 1) # + ggtitle(&#39;Consensus variants \n (PIP &gt; 0.5 in both) \n being detected&#39;)
p = p + th2 + theme(aspect.ratio = 1)
p</code></pre>
<p><img src="gtex_v8_finemap_subsample_files/figure-html/unnamed-chunk-28-3.png" width="672" /></p>
<pre class="r"><code>ggsave(&#39;../output/gtex-v8-pipeline-eqtl-finemapping-concensus-snp-with-gene-qc-stratified.png&#39;, p, height = 4, width = 8)</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
