<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Gene with low counts (comparing eQTL analysis with trcQTL)</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">mixqtl-pipeline</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/liangyy/mixqtl-pipeline">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Gene with low counts (comparing eQTL analysis with trcQTL)</h1>

</div>


<pre class="r"><code>rm(list = ls())
library(dplyr)
library(data.table)
library(ggplot2)
library(SilverStandardPerformance)
library(patchwork)
options(datatable.fread.datatable = FALSE)
source(&#39;https://gist.githubusercontent.com/liangyy/43912b3ecab5d10c89f9d4b2669871c9/raw/3ca651cfa53ffccb8422f432561138a46e93710f/my_ggplot_theme.R&#39;)
source(&#39;https://gist.githubusercontent.com/liangyy/e580a36154586148cca7fd4cd973f502/raw/bad4364b1700662c7086fcdea191e42f530d0e2e/zval2pval.R&#39;)
source(&#39;../code/rlib_analysis.R&#39;)
cbPalette &lt;- c(&#39;ascQTL&#39; = &quot;#999999&quot;, &#39;mixQTL&#39; = &quot;#E69F00&quot;, &#39;trcQTL&#39; = &quot;#56B4E9&quot;, &#39;eQTL&#39; = &quot;#009E73&quot;, &quot;#F0E442&quot;, &quot;#0072B2&quot;, &quot;#D55E00&quot;, &quot;#CC79A7&quot;)
datadir = &#39;/Users/yanyul/Desktop/mixqtl-pipeline-results/postprocess-mixqtl&#39;</code></pre>
<div id="about" class="section level1">
<h1><span class="header-section-number">1</span> About</h1>
<p>Here we meant to look at the power of eQTL and trcQTL on genes with low total count. To do so, we limit to the extreme cases where the gene has <span class="math inline">\(\le\)</span> 50 reads in all 670 samples. (See processing script <a href="https://github.com/liangyy/mixqtl-pipeline/blob/dev/misc_data/gen_gene_low_count.R">here</a>).</p>
<p>In total, we end up with 912 genes. To compare the power, we rely on eQTLGen as silver standard. We randomly selected 100,000 variant/gene pairs from the pairs with p-value &gt; 0.5 in eQTLGen (as negative instances). And similary, we selected 100,000 variant/gene pairs from the pairs with FDR &lt; 0.05 in eQTLGen (as positive instances). The processing scripts are <a href="https://github.com/liangyy/mixqtl-pipeline/blob/dev/misc_data/gen_eqtlgen_neg-with-gene-low_count.sh">here (negative ones)</a> and <a href="https://github.com/liangyy/mixqtl-pipeline/blob/dev/misc_data/gen_eqtlgen_pos-with-gene-low_count.sh">here (positive ones)</a>.</p>
<p>Some of the variant/gene pairs were not included in the GTEx cis-eQTL analysis (due to different definition on cis-window?). So, the final number of variant/gene pairs that go into analysis is less than 100,000.</p>
<p>To make a full comparison, we impute the zeros in the data with ones in the trcQTL calculation. And <code>trc_cutoff = 1</code> is used in <code>mixqtl</code> call which essentially corresponds to use all data. The script used is <a href="https://github.com/liangyy/mixqtl-pipeline/blob/dev/code/gtex_v8_mixqtl_low_count_impute.R">here</a>.</p>
</div>
<div id="load-data" class="section level1">
<h1><span class="header-section-number">2</span> Load data</h1>
<pre class="r"><code>ll = data.frame(
  prefix = c(&#39;Whole_Blood.allpairs&#39;, &#39;Whole_Blood-x-result-mixqtl_low_count_impute&#39;),
  method = c(&#39;eQTL&#39;, &#39;trcQTL&#39;),
  bcol = c(8, 4),  # c(8, 13)
  scol = c(9, 5),  # c(9, 14) 
  vcol = c(2, 16),
  gcol = c(1, 17)
)
out = list()
for(p in 1 : nrow(ll)) {
  pre = ll$prefix[p]
  b = ll$bcol[p]
  s = ll$scol[p]
  g = ll$gcol[p]
  v = ll$vcol[p]
  tag = ll$method[p]
  for(i in c(&#39;pos&#39;, &#39;neg&#39;)) {
    tmp_ = fread(cmd = paste0(&#39;zcat &lt; &#39;, datadir, &#39;/&#39;, pre, &#39;-IN-eqtlgen_&#39;, i, &#39;_low_count.subsample100000-with-gene-low-expr.txt.gz&#39;), header = TRUE)[, c(b, s, g, v)] %&gt;% mutate(method = tag, type = i) 
    colnames(tmp_)[1:4] = c(&#39;bhat&#39;, &#39;se&#39;, &#39;gene&#39;, &#39;variant&#39;)
    tmp_$gene = trim_dot(tmp_$gene)
    out[[length(out) + 1]] = tmp_
  }
}
out = do.call(rbind, out)
pairs_with_two_obs = out %&gt;% group_by(gene, variant) %&gt;% summarize(n = n()) %&gt;% filter(n == 2) %&gt;% select(gene, variant)
out = out %&gt;% filter(paste(gene, variant) %in% paste(pairs_with_two_obs$gene, pairs_with_two_obs$variant))
out = out %&gt;% mutate(stat = bhat / se)
out %&gt;% group_by(method, type) %&gt;% summarize(median(stat^2))</code></pre>
<pre><code>## # A tibble: 4 x 3
## # Groups:   method [2]
##   method type  `median(stat^2)`
##   &lt;fct&gt;  &lt;chr&gt;            &lt;dbl&gt;
## 1 eQTL   neg              0.457
## 2 eQTL   pos              3.23 
## 3 trcQTL neg              0.481
## 4 trcQTL pos              3.30</code></pre>
</div>
<div id="plot-rocpr-curves" class="section level1">
<h1><span class="header-section-number">3</span> Plot ROC/PR curves</h1>
<pre class="r"><code>curve = list()
curve2 = list()
for(m in unique(out$method)) {
  df = out %&gt;% filter(method == m)
  trues = df %&gt;% filter(type == &#39;pos&#39;) %&gt;% mutate(pair = paste(gene, variant)) %&gt;% pull(pair)
  alls = df %&gt;% mutate(pair = paste(gene, variant)) %&gt;% pull(pair)
  scores = df %&gt;% mutate(score = stat ^ 2) %&gt;% pull(score)
  o = gen_roc_curve(
    trues,
    alls, 
    scores,
    cutoff = quantile(scores, probs = 0:100 / 100)
  ) %&gt;% mutate(method = m)
  o2 = gen_fdr_power_curve(
    trues,
    alls, 
    scores,
    cutoff = quantile(scores, probs = 0:100 / 100)
  ) %&gt;% mutate(method = m)
  curve[[length(curve) + 1]] = o
  curve2[[length(curve2) + 1]] = o2
   # %&gt;% ggplot() + geom_path(aes(x = fpr, y = tpr))
}
p1 = do.call(rbind, curve) %&gt;% ggplot() + 
  geom_path(aes(x = fpr, y = tpr, color = method), size = 2, alpha = 0.5) + 
  scale_color_manual(values = cbPalette) + 
  geom_abline(intercept = 0, slope = 1, color = &#39;lightgray&#39;) +
  th + xlab(&#39;False positive rate&#39;) + ylab(&#39;True positive rate&#39;) + 
  theme(legend.position = c(0.8, 0.2), legend.title = element_blank(), legend.text = element_text(size = 18)) +
  coord_equal()
p2 = do.call(rbind, curve2) %&gt;% filter(precision != 0 &amp; recall != 0) %&gt;% ggplot() + 
  geom_path(aes(x = recall, y = precision, color = method), size = 2, alpha = 0.5) + 
  scale_color_manual(values = cbPalette) + theme(aspect.ratio = 1) +
  xlab(&#39;Recall&#39;) + ylab(&#39;Precision&#39;) + 
  theme(legend.position = c(0.3, 0.2), legend.title = element_blank(), legend.text = element_text(size = 18)) +
  th
p1</code></pre>
<p><img src="low_count_genes_files/figure-html/roc-pr-1.png" width="384" /></p>
<pre class="r"><code>p2</code></pre>
<p><img src="low_count_genes_files/figure-html/roc-pr-2.png" width="384" /></p>
<pre class="r"><code>ggsave(&#39;../output/low_count_genes_ROC.png&#39;, p1, width = 4, height = 4)
ggsave(&#39;../output/low_count_genes_PR.png&#39;, p2, width = 4, height = 4)</code></pre>
<!-- ```{r qq} -->
<!-- out_neg = out %>% filter(type == 'neg') %>% mutate(pval = zval2pval(stat)) -->
<!-- my_qqplot(out_neg$pval, out_neg$method)  -->
<!-- ``` -->
</div>
<div id="compare-test-statistic" class="section level1">
<h1><span class="header-section-number">4</span> Compare test statistic</h1>
<pre class="r"><code>recode = function(x) {
  x = as.character(x)
  o = rep(&#39;negative&#39;, length(x))
  o[x == &#39;pos&#39;] = &#39;positive&#39;
  return(o)
}
p = out %&gt;% reshape2::dcast(gene + variant + type ~ method, value.var = &#39;stat&#39;) %&gt;% mutate(type = recode(type)) %&gt;% ggplot() + geom_point(aes(x = eQTL, y = trcQTL), alpha = 0.2) + facet_wrap(~type) + th2 + geom_abline(slope = 1, intercept = 0, color = &#39;gray&#39;) + xlab(&#39;eQTL test statistic&#39;) + ylab(&#39;trcQTL test statistic&#39;) + coord_equal()
p</code></pre>
<p><img src="low_count_genes_files/figure-html/test-stat-1.png" width="480" /></p>
<pre class="r"><code>ggsave(&#39;../output/low_count_genes_test_stat.png&#39;, p, height = 4, width = 8)</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
